### ISSUE TITLE
fehlende Angebote in der Drehbuchagentur

### ISSUE THREAD

**Issue by lonk0123** [TVTower/TVTower#890]
Eigentlich sollten 4 Angebote auf dem Tisch erscheinen.

[Drehbuchagentur frischt nicht auf.zip](https://github.com/TVTower/TVTower/files/10568034/Drehbuchagentur.frischt.nicht.auf.zip)
in .zst umbennen

**Comment by GWRon** [comment]
Kann das hier noch nicht nachvollziehen:
- Einladen des Speicherstands
- ich stehe vorm Drehbuchbuero
- gehe rein: volles Angebot des Haendlers
- Kaufe eines, gehe raus
- gehe rein: volles Angebot des Haendlers

![Peek 2023-02-02 13-30](https://user-images.githubusercontent.com/2625226/216325614-f3d67dde-5da3-45ed-8121-d4aecdf8389c.gif)

**Comment by lonk0123** [comment]
Hmm, bei mir wird im Spielstand den ganzen Tag lang nichts hinzugefügt (es bleiben immer nur 4 Angebote), egal wie oft ich den Raum betrete bzw. wie viel Zeit ich verstreichen lasse.
Vielleicht ein Problem mit Scr0llbaers DB?

**Comment by nittka** [comment]
Macht es einen Unterschied, ob Du das Spiel beendest, neu startest und dann nochmal lädst?

**Comment by GWRon** [comment]
Es passiert auch nix, wenn du im "Dev-Screen" (Tab) auf Script Agency gehst - und dort Refill Orders anklickst.

**Comment by lonk0123** [comment]
> Macht es einen Unterschied, ob Du das Spiel beendest, neu startest und dann nochmal lädst?

Ja, dann funzt es. Solange ich den Spielstand aber laufen ließ, wurde nichts nachgefüllt.

> Es passiert auch nix, wenn du im "Dev-Screen" (Tab) auf Script Agency gehst - und dort Refill Orders anklickst?

Probiere ich das nächste Mal aus, wenn es passiert.

Es sollte übrigens ein ähnlicher Bug sein, zu demjenigen, der das Nicht-Nachfüllen in der Ad-Agency vor Kurzem im Forum gemeldet hatte....

**Comment by nittka** [comment]
Ich gehe davon aus, dass dieses Problem sich nicht von allein erledigt hat und noch immer auftritt!?

**Comment by lonk0123** [comment]
Ich hab den Fehler bisher nur ein einziges Mal gehabt und alle Versuche ihn zu reproduzieren, scheiterten.
Falls er wieder auftreten sollte, weiß ich zumindest, was zu tun ist bzw. kann mit Ron eine Session zur Fehlerbehebung/-analyse einlegen.
Ich werde definitiv zum nächsten größeren Update mit Balancing wieder einen neuen Spielstand starten...

**Comment by lonk0123** [comment]
Der Debug-Screen der Drehbuchagentur scheint generell nicht so zu funktionieren wie z.B. bei dem Filmhandler, da er im Gegensatz zu diesem oder dem Werbemmakler-Bildschirm bei mir nach Laden eines Spielstandes und fortlaufender Zeit immer komplett leer ist und erst nach einem manuellen Replace der Angebote der Drehbuchagentur überhaupt irgendetwas anzeigt.
Außerdem läuft da ständig ein 15-Minuten-Offer-Refill-Countdown runter, bei dem ich nicht weiß, was er für eine Funktion hat.

Ist aber noch ne alte Version von Anfang Februar...

**Comment by GWRon** [comment]
> Ich habe bei aktuellen Tests folgendes halbreproduzierbar beobachtet.
> 
> * Spiel starten
> * Debug-Screen Drehbuchagentur aufmachen
> * Drehbuchagentur betreten
> * manchmal werden im Debugscreen jetzt schon nur 4 statt 5 Titel angezeigt
> * wiederholen: Drehbuch in Koffer legen, rausgehen, wieder reingehen
> * bei genügend häufiger Wiederholung wird zwar die Drehbuchagentur wieder aufgefüllt, aber im Debugscreen werden nicht alle Titel angezeigt
> 
> Ich weiß nicht, ob das ein Bug im Debug-Screen ist, oder ob es ein Symptom eines Bugs ist, der dazu führt dass Slots und dazugehörige Daten divergieren.

ich hab das jetzt hier 2x5 (zwei komplette Spielstarts) Minuten probiert ... Geld ercheatet und dann "d" zum Drehbuch holen, raus aus Agentur, rein in die Agentur ...
immer wurde korrekt aufgefuellt. Wenn mein Koffer voll war: alles beim "Dude" abgegeben - und von neuem angefangen.

**Comment by GWRon** [comment]
```Blitzmax
Type TScriptCollection Extends TGameObjectCollection
	'stores "languagecode::name"=>"used by id" connections
	Field protectedTitles:TStringMap = New TStringMap
	'=== CACHE ===
	'cache for faster access

	'holding used scripts
	Field _usedScripts:TList = CreateList() {nosave}
	Field _availableScripts:TList = CreateList() {nosave}
	Field _parentScripts:TList = CreateList() {nosave}
```
Das ist der Fehler, der beim Einladen von @lonk0123 s  Spielstand den Debugbereich "leer" laesst.

Beim Einladen werden ja neue Objekte angelegt und dann die Informationen dort reindeserialisiert. Und da wir hier "CreateList()" stehen haben, wird das bei jedem neuen `TScriptCollection` automatisch gleich angelegt.

Und Aufrufe wie "GetUsedScriptList()" geben eben _usedScripts zurueck, wenn sie existieren.


Allerdings sollte das nicht @nittka s Fehler begruenden.

**Comment by GWRon** [comment]
OK ... habe TVTower auf den master-Stand von heute gebracht - also der, der "[Scriptagency] change refill strategy" enthaelt.

Und damit habe ich dann die 4 Drehbuecher, die @nittka beschrieben hat.

**Comment by GWRon** [comment]
@nittka 
![image](https://user-images.githubusercontent.com/2625226/224353424-cf62403b-6e20-414a-8953-e3309a8cfdd6.png)

Diese Zeile sorgt direkt nach dem Hinzufuegen der neuen Drehbuecher fuer sporadisches Entfernen eben dieser.

**Comment by GWRon** [comment]
Dachte zwar ich hatte vorhin den Fehler auch ohne den DebugStop - aber hier zeigt sich, dass im Zweifel "kein passendes" (also ein ungewolltes "newGoodScript") Drehbuch fand - und das entfernte Element in die Liste kommt.
![image](https://user-images.githubusercontent.com/2625226/224361955-82d08615-1c83-4c41-a632-73a10ea1f34d.png)

**Comment by GWRon** [comment]
Mit dem vorgeschlagenen Fix sollten die "Caches" nun nicht mehr im Wege stehen und die "good script"-Praevention muesste nun ohne den Fehler auskommen.

Ich lass hier den Bericht aber noch offen und warte auf Feedback von Euch.

**Comment by nittka** [comment]
Das Debug-Screen-Problem sollte nun behoben sein. Das Problem, dass in der Werbeagantur selbst Angebote fehlen, ist weiter offen.
